File: src/data/o1/Find_Edges_in_Shortest_Paths/Find_Edges_in_Shortest_Paths.java
import java.util.*;

class Solution {
    class Edge {
        int to;
        int weight;
        int index;

        Edge(int to, int weight, int index) {
            this.to = to;
            this.weight = weight;
            this.index = index;
        }
    }

    public boolean[] findAnswer(int n, int[][] edges) {
        // Initialize adjacency list
        List<List<Edge>> adj = new ArrayList<>();
        for(int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }

        // Build the graph and keep track of edge indices
        for(int i = 0; i < edges.length; i++) {
            int u = edges[i][0];
            int v = edges[i][1];
            int w = edges[i][2];
            adj.get(u).add(new Edge(v, w, i));
            adj.get(v).add(new Edge(u, w, i));
        }

        // Compute shortest distances from source (0)
        long[] distFromSource = dijkstra(n, adj, 0);

        // Compute shortest distances from destination (n-1)
        long[] distFromDest = dijkstra(n, adj, n-1);

        long shortestPath = distFromSource[n-1];
        boolean[] answer = new boolean[edges.length];

        // If there's no path from 0 to n-1, all edges remain false
        if(shortestPath == Long.MAX_VALUE) {
            return answer;
        }

        // Check each edge
        for(int i = 0; i < edges.length; i++) {
            int u = edges[i][0];
            int v = edges[i][1];
            int w = edges[i][2];

            // Check if this edge is part of any shortest path
            if((distFromSource[u] + w + distFromDest[v] == shortestPath) ||
               (distFromSource[v] + w + distFromDest[u] == shortestPath)) {
                answer[i] = true;
            }
        }

        return answer;
    }

    private long[] dijkstra(int n, List<List<Edge>> adj, int start) {
        long[] dist = new long[n];
        Arrays.fill(dist, Long.MAX_VALUE);
        dist[start] = 0;

        PriorityQueue<long[]> pq = new PriorityQueue<>(Comparator.comparingLong(a -> a[1]));
        pq.offer(new long[]{start, 0});

        boolean[] visited = new boolean[n];

        while(!pq.isEmpty()) {
            long[] current = pq.poll();
            int u = (int)current[0];
            long currentDist = current[1];

            if(visited[u]) continue;
            visited[u] = true;

            for(Edge edge : adj.get(u)) {
                int v = edge.to;
                int w = edge.weight;
                if(!visited[v] && currentDist + w < dist[v]) {
                    dist[v] = currentDist + w;
                    pq.offer(new long[]{v, dist[v]});
                }
            }
        }

        return dist;
    }
}

public class Find_Edges_in_Shortest_Paths {
public static void runTests(Solution solution) {
    // Test case 1: Simple connected graph with unique shortest path
    int n1 = 4;
    int[][] edges1 = {{0, 1, 1}, {1, 2, 1}, {2, 3, 1}, {0, 3, 4}};
    boolean[] expected_result1 = {true, true, true, false};
    assert Arrays.equals(solution.findAnswer(n1, edges1), expected_result1) : "Test case 1 failed. Expected " + Arrays.toString(expected_result1);

    // Test case 2: Multiple shortest paths
    int n2 = 3;
    int[][] edges2 = {{0, 1, 1}, {1, 2, 1}, {0, 2, 2}};
    boolean[] expected_result2 = {true, true, true};
    assert Arrays.equals(solution.findAnswer(n2, edges2), expected_result2) : "Test case 2 failed. Expected " + Arrays.toString(expected_result2);

    // Test case 3: Disconnected graph
    int n3 = 4;
    int[][] edges3 = {{0, 1, 1}, {2, 3, 1}};
    boolean[] expected_result3 = {false, false};
    assert Arrays.equals(solution.findAnswer(n3, edges3), expected_result3) : "Test case 3 failed. Expected " + Arrays.toString(expected_result3);

    // Test case 4: Graph with loop back to start
    int n4 = 6;
    int[][] edges4 = {{0, 1, 4}, {0, 2, 1}, {1, 3, 2}, {1, 4, 3}, {1, 5, 1}, {2, 3, 1}, {3, 5, 3}, {4, 5, 2}};
    boolean[] expected_result4 = {true, true, true, false, true, true, true, false};
    assert Arrays.equals(solution.findAnswer(n4, edges4), expected_result4) : "Test case 4 failed. Expected " + Arrays.toString(expected_result4);

    // Test case 5: Multiple paths, some edges irrelevant to shortest path
    int n5 = 5;
    int[][] edges5 = {{0, 1, 2}, {0, 2, 4}, {1, 3, 1}, {2, 3, 1}, {3, 4, 3}};
    boolean[] expected_result5 = {true, false, true, false, true};
    assert Arrays.equals(solution.findAnswer(n5, edges5), expected_result5) : "Test case 5 failed. Expected " + Arrays.toString(expected_result5);

    // Test case 6: Graph with a direct shortest path
    int n6 = 4;
    int[][] edges6 = {{2, 0, 1}, {0, 1, 1}, {0, 3, 4}, {3, 2, 2}};
    boolean[] expected_result6 = {true, false, false, true};
    assert Arrays.equals(solution.findAnswer(n6, edges6), expected_result6) : "Test case 6 failed. Expected " + Arrays.toString(expected_result6);
}

    public static void main(String[] args) {
        Solution solution = new Solution();
        for (int i = 0; i < 10000000; i++) {
            runTests(solution); // Ensure your test cases are implemented inside runTests
        }
    }
}

    

File: src/data/o1/Find_Edges_in_Shortest_Paths/Find_Edges_in_Shortest_Paths.cpp
#include <vector>
#include <queue>
#include <algorithm>
#include <climits>

using namespace std;

class Solution {
public:
    vector<bool> findAnswer(int n, vector<vector<int>>& edges) {
        // Build adjacency list: node -> list of (neighbor, weight, edge_index)
        vector<vector<pair<int, pair<int, int>>>> adj(n, vector<pair<int, pair<int, int>>>());
        int m = edges.size();
        for(int i = 0; i < m; ++i){
            int a = edges[i][0];
            int b = edges[i][1];
            int w = edges[i][2];
            adj[a].emplace_back(b, make_pair(w, i));
            adj[b].emplace_back(a, make_pair(w, i));
        }
        
        // Function to perform Dijkstra and return distance vector
        auto dijkstra = [&](int start) -> vector<long long> {
            vector<long long> dist(n, LLONG_MAX);
            priority_queue<pair<long long, int>, vector<pair<long long, int>>, std::greater<pair<long long, int>>> pq;
            dist[start] = 0;
            pq.emplace(0, start);
            
            while(!pq.empty()){
                auto [current_dist, u] = pq.top();
                pq.pop();
                
                if(current_dist > dist[u]) continue;
                
                for(auto &[v, pw] : adj[u]){
                    int w = pw.first;
                    if(dist[v] > dist[u] + w){
                        dist[v] = dist[u] + w;
                        pq.emplace(dist[v], v);
                    }
                }
            }
            return dist;
        };
        
        // Compute distances from start (0) and end (n-1)
        vector<long long> dist_start = dijkstra(0);
        vector<long long> dist_end = dijkstra(n-1);
        
        // Initialize answer vector
        vector<bool> answer(m, false);
        
        long long shortest_path = dist_start[n-1];
        if(shortest_path == LLONG_MAX){
            // No path exists from 0 to n-1, all answers remain false
            return answer;
        }
        
        // Check each edge
        for(int i = 0; i < m; ++i){
            int a = edges[i][0];
            int b = edges[i][1];
            int w = edges[i][2];
            
            // Check if edge a->b is on any shortest path
            if(dist_start[a] + (long long)w + dist_end[b] == shortest_path ||
               dist_start[b] + (long long)w + dist_end[a] == shortest_path){
                answer[i] = true;
            }
        }
        
        return answer;
    }
};

//  Test cases for the C++ solution
#include <cassert>

// Test cases for the C++ solution
void testSolution(Solution& solution) {
    // Test case 1: Simple connected graph with unique shortest path
    int n = 4;
    std::vector<std::vector<int>> edges = {{0, 1, 1}, {1, 2, 1}, {2, 3, 1}, {0, 3, 4}};
    std::vector<bool> expected_result = {true, true, true, false};  // Path: 0 -> 1 -> 2 -> 3
    assert(solution.findAnswer(n, edges) == expected_result && "Test case 1 failed.");

    // Test case 2: Multiple shortest paths
    n = 3;
    edges = {{0, 1, 1}, {1, 2, 1}, {0, 2, 2}};
    expected_result = {true, true, true};  // Paths: 0 -> 1 -> 2 and 0 -> 2
    assert(solution.findAnswer(n, edges) == expected_result && "Test case 2 failed.");

    // Test case 3: Disconnected graph
    n = 4;
    edges = {{0, 1, 1}, {2, 3, 1}};
    expected_result = {false, false};  // No path from 0 to 3
    assert(solution.findAnswer(n, edges) == expected_result && "Test case 3 failed.");

    // Test case 4: Graph with loop back to start
    n = 6;
    edges = {{0, 1, 4}, {0, 2, 1}, {1, 3, 2}, {1, 4, 3}, {1, 5, 1}, {2, 3, 1}, {3, 5, 3}, {4, 5, 2}};
    expected_result = {true, true, true, false, true, true, true, false};  // Path: 0 -> 1 -> 2 -> 3
    assert(solution.findAnswer(n, edges) == expected_result && "Test case 4 failed.");

    // Test case 5: Multiple paths, some edges irrelevant to shortest path
    n = 5;
    edges = {{0, 1, 2}, {0, 2, 4}, {1, 3, 1}, {2, 3, 1}, {3, 4, 3}};
    expected_result = {true, false, true, false, true};  // Path: 0 -> 1 -> 3 -> 4
    assert(solution.findAnswer(n, edges) == expected_result && "Test case 5 failed.");

    // Test case 6: Graph with a direct shortest path
    n = 4;
    edges = {{2, 0, 1}, {0, 1, 1}, {0, 3, 4}, {3, 2, 2}};
    expected_result = {true, false, false, true};  // Path: 0 -> 1 -> 3
    assert(solution.findAnswer(n, edges) == expected_result && "Test case 6 failed.");
}


int main() {
    Solution solution;
    for (int i = 0; i < 163840; ++i) {
        testSolution(solution);  
    }
    return 0;
}
